src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -267,9 +267,10 @@ ASR::TranslationUnit_t* compile_module_till_asr(Allocator& al,
     // save_pyc_files(*r2.result, infile + "c");
     std::string input;
     read_file(infile, input);
+    lm.init_simple(input);
     std::cerr << diagnostics.render(input, lm, compiler_options);
     if (!r2.ok) {
-        LFORTRAN_ASSERT(diagnostics.has_error())
+        // LFORTRAN_ASSERT(diagnostics.has_error())
```

Comment:
I think you have to keep this in there. This checks that the diagnostics contain an error message if `r2` returned an Error.

Reply:
It seems one of the possible use could be `diagnostics` just containing a note/warning and still `throw SemanticAbort()` being called to stop further processing of `AST->ASR`.  

For example, for the below message there is no error (there is a note only), but the processing of the contents of `b.y` still needs to be stopped because we could not import `a.py`.
```bash
note: Imported here
 --> /home/ubaid/OpenSource/lpython/src/bin/../runtime/b.py:1:1
  |
1 | from a import i
  | ^^^^^^^^^^^^^^^ 
```

---

>I think you have to keep this in there. This checks that the diagnostics contain an error message if r2 returned an Error.

It seems `throw SemanticAbort()` can halt the current processing of `AST->ASR`, and on using it, due the ` LFORTRAN_ASSERT(diagnostics.has_error())`,  it is being compulsory that `diagnostics` would be containing an error. Although, in the above `b.py` example, it is not containing any error, instead it is containing a note.

Please, could someone possibly share if there is a way to stop processing of `AST->ASR` apart from using `throw SemanticAbort()`?

Reply:
I guess first, `from b import j` will happen, then `from a import i` and then finally, `i: i32 = 3.142`. So when you get the error in `a.py`, diagnostics should know that `from b import j` and then `from a import i` before `i: i32 = 3.142` have happened. More like a stack of diagnostics. Fill the diagnostic with a note when a module import happens because you know you are going to jump into external file. If the external file processes successfully then pop the note from the diagnostics. If an error happens then you already have the information of all the imports that have happened before this error and you can pop them and show to the user.

Reply:
I agree, with this approach! @Shaikh-Ubaid give it a try.
Any doubts please ask.

Reply:
Here's what I am imagining of the control flow that happens,

1. Control is in `c.py` and you processed, `from b import j`. The diagnostics stack (abstract concept, upto you how you implement it) contains that `b` was imported. A recursive call to `python_ast_to_asr` is made for processing `b.py`.
2. Now control jumped to `b.py`. Here it processed, `from a import i`. The diagnostics stack adds that `a` was imported. A recursive call to `python_ast_to_asr` is made for processing `a.py`.
3. Control is in `a.py`. Now, `i: i32 = 3.142` resulting in a semantic error. Its at this point we first add this error to diagnostics stack and then while rendering diagnostics stack keep popping and displaying to the user. And halt the processing/compilation. No need to go back to `b.py`.

I don't know about the implementation details but conceptually and ideally IMO that's how it should be done. Whenever you jump to another file, i.e., make a recursive call to `python_ast_to_asr` just store the diagnostics of import statement (which is making you to jump to another file which is being imported) in a conceptual stack and move ahead. If you face an error somewhere deep in a file, you already know from the conceptual stack how you reached there and you can use that info along with current semantic error to show the complete trace.

Make sense to you folks @Shaikh-Ubaid, @certik, @Thirumalai-Shaktivel?

Reply:
Plus I will look at your later today to understand the details of your issue better. Meanwhile may be just try removing the duplicated note I mentioned about in https://github.com/lcompilers/lpython/pull/1294#issuecomment-1313988249.